(PUP-8232) Add back 'puppet facts upload' face and update

Sharpie opened the pull request:
This changeset reverts pieces of [PUP-2560](https://tickets.puppetlabs.com/browse/PUP-2560) in order to re-add the ability to upload facts outside of a catalog request. Specifically, this patch re-adds the following functionality:

  - The `puppet facts upload` command.

  - The Puppet::Node::Facts::Rest terminus for saving facts to Puppet Servers.

  - The ability to handle requests for the `/puppet/v3/facts` endpoint.

Sharpie commented:
Once this is merged, puppetlabs/puppetserver#1578 can be merged to add the `/puppet/v3/facts` route to the Puppet Server API.

puppetcla commented:
CLA signed by all contributors.

Sharpie commented:
Oddly the style check for commit messages was fine with (SERVER-) being used in a6dd8b9 but not cf51e04...

justinstoller commented:
Alright, I'm going to be that guy. Despite everything old being new again, I don't think we should be adding indirection termini or server-side routes in Ruby in 2017 (or waiting and doing it in 2018, either).

The approach I would prefer to see taken is the the upload command making an HTTP PUT (not using the indirector) to a route not in the /v3 routes, with that corresponding new route implement on the server side in clojure (not delegating to Ruby).

That might not be feasible for reasons, but I'd like to discuss those reasons or at least have them articulated here and/or in the ticket for posterity's sake.

Sharpie commented:
For me, the main reasons come down to avoiding code duplication and creating divergent methods for serializing data between wire formats. Additionally, the goal is to replicate the "fact update" effects of a catalog request without the overhead of catalog compilation, and the relevant code pathways from a catalog request involve using the indirector.

PuppetDB and Puppet Server both have different wire formats for fact data:

  - https://github.com/puppetlabs/puppet/blob/5.3.3/api/schemas/facts.json
  - https://puppet.com/docs/puppetdb/5.1/api/wire_format/facts_format_v5.html

The server-side library for translating between the two formats is part of the terminus code in the PuppetDB project along with other helper code related to forming and profiling PuppetDB API requests:

 https://github.com/puppetlabs/puppetdb/blob/5.1.3/puppet/lib/puppet/indirector/facts/puppetdb.rb#L19-L54

On the agent side, much of the logic around converting the `Puppet::Node::Facts` object returned by Facter to the appropriate HTTP call and payload format lives in `Puppet::Indirector::REST` along with logic for adding expected headers and selecting the right server in a HA configuration.

The major benefit from shifting this operation over to Clojure would be reducing load on the JRuby worker pool. I would expect this to provide significantly better performance in an environment where a lot of nodes were generating fact upload activity.

However, on the con side we would have to duplicate the agent and server side behaviors listed above and ensure parity is maintained where needed on an ongoing basis. Additionally, it is currently possible for folks to run Puppet Server with something other than PuppetDB as the facts cache by configuring `routes.yaml` to used a different indirector terminus. Implementing this behavior solely in Clojure wouldn't work for those use cases. Finally, there is a secondary benefit to having an API for independently exercising the same code paths used to persist facts during catalog compilation. This  is useful in certain debugging scenarios as a method for dividing the problem space.

So, there is the potential for significant performance gains from moving this into Clojure, but at the cost of an initial re-write plus ongoing maintenance and potential inconsistency compared with how fact caching behaves during catalog requests. On the balance I think this weighs out in favor of re-using the indirector framework to submit facts.

justinstoller commented:
Thank you for the thoughtful reply, Charlie.

As much as I don't like seeing us add to the indirector, unless we're going to prioritize a larger shift away from it, this is probably the right approach.

joshcooper commented on lib/puppet/face/facts.rb:
```diff
@@ -35,4 +35,43 @@
+      # etc. instead of `user` section settings.
+      Puppet.settings.preferred_run_mode = :agent
+      Puppet::Node::Facts.indirection.terminus_class = :facter
+      facts = Puppet::Node::Facts.indirection.find(Puppet[:certname])
```
I think we want to use `Puppet[:node_name_value]` which defaults to `$certname` but it doesn't have to be the same. Also `node_name_value` is what the agent uses when requesting node & catalog and sending reports.

Sharpie replied:
Hmm, seems like a good idea --- but I can't recall off the top of my head whether `node_name_value` is dependent on `node_name_fact`, which requires loading facts, which puts us in a bit of a Catch-22. Also, the node name used for the lookup might not even matter when the terminus is set to `:facter`.

Definitely seems like the thing to use when POSTing facts back to the server though. Maybe just need to ensure the `name` attribute of the returned factset is populated before the save call.

joshcooper replied:
`node_name_value` and `node_name_fact` are mutually exclusive. Using the fact-based one definitely has problems. You're right that the name we request does not matter when using the `Facter` terminus, except that it becomes the `name` attribute of the returned facts object. So I think we want to use `node_name_value` so that the returned facts object is consistent with what puppet thinks it's name is.

Sharpie replied:
Looks like the common pattern used by `apply` and agent is to query using `Puppet[:node_name_value]` and then re-set `Puppet[:node_name_value]` and the `name` attribute of the returned factset if `Puppet[:node_name_fact]` is not empty:

https://github.com/puppetlabs/puppet/blob/5.3.4/lib/puppet/application/apply.rb#L204-L212
https://github.com/puppetlabs/puppet/blob/5.3.4/lib/puppet/configurer/fact_handler.rb#L15-L19

I'll implement the strategy from `fact_handler.rb` and add some test coverage.

Sharpie commented:
Updated `puppet facts upload` to use `node_name_value` and `node_name_fact` instead of `certname`.

ericldelaney commented:
```
$ sum ~/.puppetlabs/opt/puppet/cache/yaml/facts/localhost.yaml
25923 4 /Users/eric.delaney/.puppetlabs/opt/puppet/cache/yaml/facts/localhost.yaml
$ bx puppet agent -t
Info: Using configured environment 'production'
Info: Retrieving pluginfacts
Info: Retrieving plugin
Info: Retrieving locales
Info: Caching catalog for localhost
Info: Applying configuration version '1519415605'
Notice: Applied catalog in 0.02 seconds
$ sum ~/.puppetlabs/opt/puppet/cache/yaml/facts/localhost.yaml
28996 4 /Users/eric.delaney/.puppetlabs/opt/puppet/cache/yaml/facts/localhost.yaml
$
$ bx puppet agent -t --use_cached_catalog
Info: Using cached catalog from environment 'production'
Info: Applying configuration version '1519415605'
Notice: Applied catalog in 0.02 seconds
$ sum ~/.puppetlabs/opt/puppet/cache/yaml/facts/localhost.yaml
28996 4 /Users/eric.delaney/.puppetlabs/opt/puppet/cache/yaml/facts/localhost.yaml
$
$ bx puppet facts upload
Notice: Uploading facts for 'localhost' to: 'localhost'
$ sum ~/.puppetlabs/opt/puppet/cache/yaml/facts/localhost.yaml
28719 4 /Users/eric.delaney/.puppetlabs/opt/puppet/cache/yaml/facts/localhost.yaml
$
```

ericldelaney commented:
@Sharpie Could you squash the commits or change them to be (PUP-8232) so that the commit checker will be happy.

Sharpie commented:
I suppose I could re-tag the SERVER-2054 commits as PUP-8232, but the commit checker not recognizing that Puppet changes may be required to implement Puppet Server features seems like a bug in the commit checker.

ericldelaney commented:
@Sharpie the commit checker for the puppet repo is very picky it needs to be a PUP ticket or one of the allowed tags or it will complain. That's true for all commits that get added.

The SERVER ticket and PR are find, its just this one since its going into puppet

Sharpie commented:
Right. But those particular Puppet code changes are for the purpose of a implementing a server API endpoint. Those commits are for a SERVER feature, some of the code just happens to live in the Puppet repo for historical reasons.

ericldelaney commented:
@justinstoller Could you please review this, I've tested it and SERVER-2054 together and we can properly upload facts. I added the auth updates to my server running from source and was able to unable to access other nodes facts (like we want).

puppetcla commented:
CLA signed by all contributors.

joshcooper commented on lib/puppet/face/facts.rb:
```diff
@@ -35,4 +35,48 @@
+                    node: Puppet[:node_name_value],
+                    server: server})
+
+      Puppet::Node::Facts.indirection.save(facts)
```
@Sharpie can the caller try to pass trusted facts here? Does puppetserver and/or puppetdb terminus reject the request, or does it strip them out?

Sharpie replied:
That is a very good question. I'll look into what happens on the server side.

joshcooper replied:
Similar question, is the timestamp in the uploaded facts ignored? Or does the puppetdb terminus always ignore it and set the "server-specified" timestamp to avoid clock skew issues?

ericldelaney replied:
@Sharpie have you had a chance to look into Josh's questions?

Sharpie replied:
Not yet. I've been tied up for the past two days with a customer perf problem and MCollective issues :(

ericldelaney replied:
Here is the list of facts passed to the save call on line 79, we don't appear to have any of the trusted facts here:

```
#<Puppet::Node::Facts:0x007fc78e116a40 @name="localhost",
 @values={
 "puppetversion"=>"5.3.5",
 "kernelversion"=>"17.4.0",
 "uptime"=>"7 days",
 "kernelrelease"=>"17.4.0",
 "uptime_seconds"=>681530,
 "path"=>"/Users/eric.delaney/work/my_repos/puppetserver/ruby/puppet/.bundle/ruby/2.2.0/bin:/Users/eric.delaney/.rbenv/versions/2.2.8/bin:/usr/local/Cellar/rbenv/1.1.1/libexec:./ruby/puppet/bin:./ruby/facter/bin:/Users/eric.delaney/.rbenv/shims:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/opt/puppetlabs/bin:/usr/local/munki:/Users/eric.delaney/bin:/Users/eric.delaney/.rvm/bin",
 "kernel"=>"Darwin",
 "zzz_test"=>"data",
 "facter"=>".puppetlabs",
 "xxx_test"=>"thing",
 "test"=>"testing",
 "hostname"=>"MacBook-Pro-2",
 "productname"=>"MacBookPro11,4",
 "operatingsystem"=>"Darwin",
 "processors"=>{"count"=>8,
 "speed"=>"2.2 GHz"},
 "uptime_days"=>7,
 "kernelmajversion"=>"17.4",
 "timezone"=>"PST",
 "ps"=>"ps auxwww",
 "virtual"=>"physical",
 "is_virtual"=>false,
 "architecture"=>"x86_64",
 "hardwaremodel"=>"x86_64",
 "os"=>{"name"=>"Darwin",
 "family"=>"Darwin",
 "release"=>{"major"=>"17",
 "minor"=>"4",
 "full"=>"17.4.0"}},
 "network_lo0"=>"127.0.0.0",
 "network_en0"=>"10.0.24.0",
 "rubysitedir"=>"/Users/eric.delaney/.rbenv/versions/2.2.8/lib/ruby/site_ruby/2.2.0",
 "sp_smc_version_system"=>"2.29f24",
 "sp_boot_rom_version"=>"MBP114.0177.B00",
 "sp_cpu_type"=>"Intel Core i7",
 "sp_current_processor_speed"=>"2.2 GHz",
 "sp_l2_cache_core"=>"256 KB",
 "sp_l3_cache"=>"6 MB",
 "sp_machine_model"=>"MacBookPro11,4",
 "sp_machine_name"=>"MacBook Pro",
 "sp_number_processors"=>4,
 "sp_packages"=>1,
 "sp_physical_memory"=>"16 GB",
 "sp_platform_uuid"=>"EC54FA6D-EB0B-5DE3-B6B0-81D73DF53409",
 "sp_serial_number"=>"C02T50DPG8WL",
 "sp_boot_mode"=>"normal_boot",
 "sp_boot_volume"=>"Macintosh HD",
 "sp_kernel_version"=>"Darwin 17.4.0",
 "sp_local_host_name"=>"MacBook Pro (2)",
 "sp_os_version"=>"macOS 10.13.3 (17D102)",
 "sp_secure_vm"=>"secure_vm_enabled",
 "sp_system_integrity"=>"integrity_enabled",
 "sp_uptime"=>"up 7:21:19:20",
 "sp_user_name"=>"Eric Delaney (eric.delaney)",
 "macosx_productname"=>"Mac OS X",
 "macosx_productversion"=>"10.13.3",
 "macosx_buildversion"=>"17D102",
 "macosx_productversion_major"=>"10.13",
 "macosx_productversion_minor"=>"3",
 "operatingsystemmajrelease"=>"17",
 "gid"=>"staff",
 "hardwareisa"=>"i386",
 "rubyversion"=>"2.2.8",
 "processorcount"=>8,
 "system_uptime"=>{"seconds"=>681530,
 "hours"=>189,
 "days"=>7,
 "uptime"=>"7 days"},
 "macaddress"=>"18:65:90:cb:91:23",
 "osfamily"=>"Darwin",
 "id"=>"eric.delaney",
 "uptime_hours"=>189,
 "facterversion"=>"2.5.1",
 "memorysize"=>"16.00 GB",
 "memoryfree"=>"82.46 MB",
 "swapsize"=>"2.00 GB",
 "swapfree"=>"1.15 GB",
 "swapsize_mb"=>"2048.00",
 "swapfree_mb"=>"1181.00",
 "memorysize_mb"=>"16384.00",
 "memoryfree_mb"=>"82.46",
 "swapencrypted"=>true,
 "netmask"=>"255.255.252.0",
 "domain"=>"local",
 "interfaces"=>"lo0,gif0,stf0,XHC20,en0,p2p0,awdl0,en1,en2,bridge0,utun0",
 "ipaddress_lo0"=>"127.0.0.1",
 "netmask_lo0"=>"255.0.0.0",
 "mtu_lo0"=>16384, "mtu_gif0"=>1280, "mtu_stf0"=>1280, "mtu_xhc20"=>0, "ipaddress_en0"=>"10.0.26.32",
 "macaddress_en0"=>"18:65:90:cb:91:23",
 "netmask_en0"=>"255.255.252.0",
 "mtu_en0"=>1500, "macaddress_p2p0"=>"0a:65:90:cb:91:23",
 "mtu_p2p0"=>2304, "macaddress_awdl0"=>"46:e9:09:a3:de:62",
 "mtu_awdl0"=>1484, "macaddress_en1"=>"6a:00:02:fa:21:a0",
 "mtu_en1"=>1500, "macaddress_en2"=>"6a:00:02:fa:21:a1",
 "mtu_en2"=>1500, "macaddress_bridge0"=>"6a:00:02:fa:21:a0",
 "mtu_bridge0"=>1500, "mtu_utun0"=>2000, "fqdn"=>"MacBook-Pro-2.local",
 "operatingsystemrelease"=>"17.4.0",
 "ipaddress"=>"10.0.26.32",
 "rubyplatform"=>"x86_64-darwin17",
 "clientcert"=>"localhost",
 "clientversion"=>"5.3.5",
 "clientnoop"=>false
},
@timestamp=2018-02-28 12:03:53 -0800,
@expiration=2018-02-28 12:33:53 -0800>
```

Sharpie replied:
Yeah, the `save` call wouldn't receive trusted facts unless there was a custom fact that produced a value called `trusted`.

What should happen:

  - The `/puppet/v3/facts` API should ignore any `trusted` key in the `values` section of the PUT body.

  - If the client certificate used to make connection is valid and contains trusted data, this data should be merged in as an allowed `trusted` value.

I'd bet the current API endpoint fails both of those as it pre-dated trusted data.

Sharpie replied:
Ok, looks like the PuppetDB Facts terminus does exactly the right thing during a `save` action by consulting `Puppet.lookup(:trusted_information)` which the HTTP handler sets to info recovered from the client cert, if present:

https://github.com/puppetlabs/puppetdb/blob/5.2.0/puppet/lib/puppet/indirector/facts/puppetdb.rb#L28

However, the YAML terminus will just save whatever is submitted. This can cause some problems during compilation as certain Puppet Server actions will load data from the YAML cache rather than hitting PuppetDB.

joshcooper commented on lib/puppet/face/facts.rb:
```diff
@@ -35,4 +35,48 @@
+      not use this facility, and it is turned off by default. See
+      <https://puppet.com/docs/puppetserver/latest/config_file_auth.html>
+      for more details.
+    EOT
```
@gguillotte should we be referring to `puppet server` instead or some variation?

gguillotte replied:
I can't really tell from the context. Is this feature available only on Puppet Server masters, and/or are we testing or supporting this with Puppet Server only? If so, I'd specify that, like on L47:

```
      This action requires that Puppet Server's `auth.conf` file 
```

joshcooper replied:
The REST endpoint would be available in both, but we're only changing the default auth.conf in puppetserver, right @Sharpie?

Sharpie replied:
The default `auth.conf` for Puppet Server is updated over in puppetlabs/puppetserver#1578. Given that Rack and the ruby-based auth.conf are both deprecated, I think it makes more sense to update these docs to say "Puppet Server" and just leave Rack masters out of it.

kris-bosland commented:
Hi @Sharpie, with the 5.5.x branch being cut, please retarget that branch.  Thanks!

joshcooper commented:
So I think we're in good shape, just had a few final comments:

1. Adding the rest terminus for facts automagically adds support for `puppet facts find --terminus rest`. Should the default auth rules allow download too? If it's not needed I'd prefer to only allow `save` and not `find` by default.
2. Since this is landing in 5.5, I'd feel better if both webrick/passenger and puppetserver supported the same REST endpoints and auth rules. Could you update `conf/auth.conf`?
3. I think we want to allow the caller to specify the environment (such as in agent-specified environment workflows), but the `Puppet::Application::Facts` assumes the per-environment directory exists locally. We probably want to add `environment_mode :not_required` like we do for `Puppet::Application::Config`.

Sharpie commented:
> Adding the rest terminus for facts automagically adds support for puppet facts find --terminus rest. Should the default auth rules allow download too? If it's not needed I'd prefer to only allow save and not find by default.

Good point. puppetlabs/puppetserver#1578 only adds a HTTP route handler for PUT requests, so `find` wouldn't work anyway. If folks want to retrieve facts via an API, then PuppetDB is the thing to use.

> Since this is landing in 5.5, I'd feel better if both webrick/passenger and puppetserver supported the same REST endpoints and auth rules. Could you update conf/auth.conf?

👍 It's a simple change.

> I think we want to allow the caller to specify the environment (such as in agent-specified environment workflows), but the Puppet::Application::Facts assumes the per-environment directory exists locally. We probably want to add environment_mode :not_required like we do for Puppet::Application::Config.

Makes sense. I'll get that added.

Sharpie commented:
Looked into removing the `find` handler and it doesn't look like there's a good way to do this in the code. It appears that for similar non-sensical requests like `destroy` on `/puppet/v3/catalog` or `find` on `/puppet/v3/reports` we handle this situation by:

  - Not having a Ring handler that responds to that HTTP method in Puppet Server, which results in a 404 error.

  - Using the default "deny all" behavior of `auth.conf` to reject that method with a 403 Forbidden in a Rack Puppet Master.

So, we can likely handle this by only adding handlers and auth rules for save/PUT requests.

Sharpie commented:
Added a legacy `auth.conf` to keep parity with puppetlabs/puppetserver#1578 for Rack masters and updated `lib/puppet/application/facts.rb` with `environment_mode :not_required`.

joshcooper commented on lib/puppet/indirector/facts/yaml.rb:
```diff
@@ -5,10 +5,6 @@ class Puppet::Node::Facts::Yaml < Puppet::Indirector::Yaml
-  def allow_remote_requests?
-    false
-  end
-
```
I also verified yaml is still rejected due to the other checks we have in place, so 👍
